src/librustc_trans/collector.rs

```diff
@@ -456,12 +456,25 @@ impl<'a, 'tcx> MirVisitor<'tcx> for MirNeighborCollector<'a, 'tcx> {
         match *rvalue {
             mir::Rvalue::Aggregate(mir::AggregateKind::Closure(def_id,
                                                                ref substs), _) => {
-                assert!(can_have_local_instance(self.scx.tcx(), def_id));
-                let trans_item = create_fn_trans_item(self.scx.tcx(),
-                                                      def_id,
-                                                      substs.func_substs,
-                                                      self.param_substs);
-                self.output.push(trans_item);
+                let mir = errors::expect(self.scx.sess().diagnostic(), self.scx.get_mir(def_id),
+                                         || format!("Could not find MIR for closure: {:?}", def_id));
+
+                let concrete_substs = monomorphize::apply_param_substs(self.scx.tcx(),
+                                                                       self.param_substs,
+                                                                       &substs.func_substs);
+                let concrete_substs = self.scx.tcx().erase_regions(&concrete_substs);
+
+                let mut visitor = MirNeighborCollector {
```

Comment:
what's going on here? This seems pretty similar to what the old code used to do, just inlined, right? That is, the old code would push a `TransItem::Fn`, which would then do something roughly like this in `collect_items_rec`?